File: app/src/main/java/com/larryhsiao/nyx/tags/JotTagRemoval.java
package com.larryhsiao.nyx.tags;

import com.silverhetch.clotho.Action;
import com.silverhetch.clotho.Source;

import java.sql.Connection;
import java.sql.PreparedStatement;

/**
 * Action to remove attached tag from jot.
 */
public class JotTagRemoval implements Action {
    private final Source<Connection> conn;
    private final long jotId;

    public JotTagRemoval(Source<Connection> conn, long jotId) {
        this.conn = conn;
        this.jotId = jotId;
    }

    @Override
    public void fire() {
        try (PreparedStatement stmt = conn.value().prepareStatement(
            // language=H2
            "DELETE FROM tag_jot WHERE jot_id=?"
        )) {
            stmt.setLong(1, jotId);
            stmt.executeUpdate();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: core/src/main/java/com/larryhsiao/nyx/core/tags/JotTagRemoval.kt
package com.larryhsiao.nyx.core.tags

import com.larryhsiao.clotho.Action
import com.larryhsiao.clotho.Source
import java.sql.Connection

/**
 * Action to remove attached a tag from jot.
 */
class JotTagRemoval(
    private val conn: Source<Connection>,
    private val jotId: Long,
    private val tagId: Long
) : Action {
    override fun fire() {
        try {
            conn.value().prepareStatement( // language=H2
                "UPDATE tag_jot " +
                    "SET DELETE = 1,  VERSION = VERSION + 1" +
                    " WHERE jot_id=? AND TAG_ID=?"
            ).use { stmt ->
                stmt.setLong(1, jotId)
                stmt.setLong(2, tagId)
                stmt.executeUpdate()
            }
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }
    }
}